janus_server/src/datastore.rs

```diff
@@ -665,6 +673,90 @@ impl Transaction<'_> {
+        &self,
+        lease_duration: Duration,
+        maximum_acquire_count: usize,
+    ) -> Result<Vec<(TaskId, VdafInstance, AggregationJobId, Time)>, Error> {
+        let now = self.clock.now();
+        let lease_expiry_time = now.add(lease_duration)?;
+        let maximum_acquire_count: i64 = maximum_acquire_count.try_into()?;
+
+        // TODO(brandon): verify that this query is efficient. I am not sure if we would currently
+        // scan over every (in-progress, not-leased) aggregation job for tasks where we are in the
+        // HELPER role.
+        let stmt = self
+            .tx
+            .prepare_cached(
+                "UPDATE aggregation_jobs SET lease_expiry = $1
+                    FROM tasks
+                    WHERE tasks.id = aggregation_jobs.task_id
+                    AND aggregation_jobs.id IN (SELECT aggregation_jobs.id FROM aggregation_jobs
+                        JOIN tasks on tasks.id = aggregation_jobs.task_id
+                        WHERE tasks.aggregator_role = 'LEADER'
+                        AND aggregation_jobs.state = 'IN_PROGRESS'
+                        AND (aggregation_jobs.lease_expiry IS NULL
+                          OR aggregation_jobs.lease_expiry <= $2)
+                        LIMIT $3)
+                    RETURNING tasks.task_id, tasks.vdaf, aggregation_jobs.aggregation_job_id",
+            )
+            .await?;
+        self.tx
+            .query(
+                &stmt,
+                &[
+                    /* lease_expiry */ &lease_expiry_time.as_naive_date_time(),
+                    /* now */ &now.as_naive_date_time(),
+                    /* limit */ &maximum_acquire_count,
+                ],
+            )
+            .await?
+            .into_iter()
+            .map(|row| {
+                let task_id = TaskId::get_decoded(row.get("task_id"))?;
+                let vdaf = row.try_get::<_, Json<VdafInstance>>("vdaf")?.0;
+                let aggregation_job_id =
+                    AggregationJobId::get_decoded(row.get("aggregation_job_id"))?;
+                Ok((task_id, vdaf, aggregation_job_id, lease_expiry_time))
+            })
+            .collect()
+    }
+
+    /// release_aggregation_job releases an acquired (via e.g. acquire_incomplete_aggregation_jobs)
+    /// aggregation job. It returns an error if the aggregation job has no current lease.
+    pub async fn release_aggregation_job(
+        &self,
+        task_id: TaskId,
+        aggregation_job_id: AggregationJobId,
+    ) -> Result<(), Error> {
+        // TODO(brandon): sanity-check that the aggregation job is leased?
+        let stmt = self
+            .tx
+            .prepare_cached(
+                "UPDATE aggregation_jobs SET lease_expiry = NULL
+                FROM tasks
+                WHERE tasks.id = aggregation_jobs.task_id
+                  AND tasks.task_id = $1
+                  AND aggregation_jobs.aggregation_job_id = $2",
```

Comment:
There's an argument to be made for this, but I think we want to update it in another PR -- this would be a change to how we address entities currently. That is, currently, entities that live inside a task are always effectively identified by `task_id` as well as their own ID, even if that ID is likely globally unique.

In at least one case, I think keeping these semantics is a good idea: client reports are identified by `task_id` & `nonce`; since nonces are generated by clients, I think it's probably a good idea to keep client reports identified by `task_id` as well to keep different tasks isolated from one another. OTOH, this argument doesn't apply to aggregation jobs, since the ID is generated by the servers, which I think can be trusted to produce "good" random values. But consistency is likely valuable--it would be a little funky to require `task_id` & `nonce` for reports, but only `aggregation_job_id` for aggregation jobs.